### PR TITLE
References ($ref) and JSON pointers

### PR DESCRIPTION
**Pull request by ammancilla**
Currently, `blumquist` assumes that **all the references** (`$ref`) listed in a given schema **are contained in the schema** `definitions` **section**. Unfortunately, the [json-schema specification](https://cswr.github.io/JsonSchema/spec/definitions_references/#json-pointers) disagrees, as a reference could be an arbitrary **JSON pointer**, which might point to any section of a JSON document.

This  PR will allow `blumquist` to handle references in a better way, by adding support to resolve ~a reduced amount of~ different JSON pointer formats.

- Add`Blumquist::JSONPointer` to resolve JSON pointers.
- Use `Blumquist::JSONPointer` to resolve references.
- Add Support the following JSON pointer formats:

     1. `#/key1/key2/.../keyN`
     2. `path_to_file.json`
     3. `path_to_file.json#/key1/key2/.../keyN`

### CODE REVIEW

File: lib/blumquist/errors/invalid_pointer.rb [hunk 1]
```diff
@@ -1,8 +1,13 @@
 class Blumquist
   module Errors
     class InvalidPointer < Blumquist::Error
-      def initialize(pointer)
-        super("Could not find pointer #{pointer.to_s.to_json}")
+      def initialize(pointer:, document:)
+          msg = %{
+  Could not find pointer '#{pointer}' in the given document:\n
+  #{JSON.pretty_generate(document)}
+          }
```
Comment (ammancilla) [inline lib/blumquist/errors/invalid_pointer.rb]: The indentation might looks wrong, but it is not.  `%{}` take the indentation into account.

File: lib/blumquist.rb [hunk 2]
```diff
@@ -2,9 +2,12 @@
 require 'active_support/core_ext/hash/indifferent_access'
 require 'json'
 require 'json-schema'
+require 'blumquist/json_pointer'
 require 'blumquist/errors'
 
 class Blumquist
+  PRIMITIVE_TYPES = %w{null boolean number string enum integer}.freeze
```
Comment (ammancilla) [inline lib/blumquist.rb]: Moved to a constant as this list is also used [here](https://github.com/moviepilot/blumquist/pull/22/files#diff-323f52f0e9e59e9550be13a6fd29dfaeR5).

File: spec/support/phone_number_schema.json [hunk 3]
```diff
@@ -0,0 +1,13 @@
+{
```
Comment (ammancilla) [inline spec/support/phone_number_schema.json]: Moved to a different file to test JSON pointers to external files.

File: spec/support/schema.json [hunk 4]
```diff
@@ -80,21 +80,21 @@
...
 
+  "title": "Company",
   "type": "object",
-
   "properties": {
     "name": { "type": "string", "maxLength": 16 },
     "phone_numbers": {
       "type": "array",
-      "items": [{ "type": "object", "properties": { "prefix": {"type" :"number"}, "extension": {"type": "number"} }}]
+      "items": [{"$ref": "spec/support/phone_number_schema.json"}]
```
Comment (ammancilla) [inline spec/support/schema.json]: Pointer to external file 🎉

File: lib/blumquist/json_pointer.rb [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    end
+
+    def validate_uri!
+      #
+      # Why only this uri formats are valid?
+      # Refer to the description on the top of this file
+      #
+
+      return if uri.start_with?(DOCUMENT_ROOT_IDENTIFIER)
+      return if uri.match(DOCUMENT_ADDRESS_IDENTIFIER)
```
Comment (ammancilla) [inline lib/blumquist/json_pointer.rb]: I know this is not a proper way to validate the `uri` but is enough to start

File: lib/blumquist/errors/invalid_pointer.rb (hunk 1 above)
Comment (jayniz) [reply lib/blumquist/errors/invalid_pointer.rb]: how about 

```ruby
    msg = "foo " \
                "bar"
```

or 

```ruby
   msg = <<~MSG
     foo bar
MSG
```

Comment (ammancilla) [reply lib/blumquist/errors/invalid_pointer.rb]: The second one is equivalent to the one used here. The first one might work better, although I don't think it is an issue to use `%{}` and have the indentation like that.

**Pull request merged**